packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,152 @@
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+import autobind from 'autobind-decorator'
+import themed from 'react-themed'
+import { EVENTS } from './utils/const'
+import { setupEvents, removeEvent } from './utils/mapEventHelpers'
+
+@themed(['Gmap_FreeDrawTootip'], { pure: true })
+export default class FreeDrawLayer extends PureComponent {
+  static propTypes = {
+    map: PropTypes.object,
+    onMapDrawStart: PropTypes.func,
+    onMapDrawEnd: PropTypes.func,
+    theme: PropTypes.object,
+    mapControls: PropTypes.object.isRequired,
+    shapes: PropTypes.Array,
+  }
+
+  static defaultProps = {
+    theme: {},
+    map: {},
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      polygons: [],
+    }
+  }
+
+  enableMapControls() {
+    const { map, mapControls } = this.props
+    map.setOptions({
+      draggable: true,
+      zoomControl: true,
+      scrollwheel: true,
+      disableDoubleClickZoom: true,
+      ...mapControls,
+    })
+  }
+
+  disableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: false,
+      zoomControl: false,
+      scrollwheel: false,
+      disableDoubleClickZoom: false,
+    })
+  }
+
+  @autobind
+  enableMapDraw() {
+    const { map } = this.props
+    this.disableMapControls()
+    setupEvents(map, EVENTS, {
+      onMouseDown: () => {
+        this.drawFreeHand()
+      },
+    }, true)
+  }
+
+  drawFreeHand() {
+    const { map, onMapDrawStart } = this.props
+    const poly = new window.google.maps.Polyline({ map,
+      clickable: false,
+      strokeColor: '#d32526',
+      strokeOpacity: 1,
+      strokeWeight: 2,
+    })
+
+    const drawPoly = new window.google.maps.Polygon({
+      map,
+      path: poly.getPath(),
+      clickable: false,
+      strokeColor: '#d32526',
+      strokeOpacity: 0.8,
+      strokeWeight: 0,
+      fillColor: '#9b9b9b',
+      fillOpacity: 0.30,
+    })
+
+    if (onMapDrawStart) {
+      onMapDrawStart()
+    }
+    const listener = setupEvents(map, EVENTS, {
+      onMouseMove: e => {
+        poly.getPath().push(e.latLng)
+      },
+    }, false)
+
+    setupEvents(map, EVENTS, {
+      onMouseUp: () => {
+        removeEvent(listener.onMouseMove)
```

Comment:
Do we need this anymore since you're setting up the event as a one time listener now?

Reply:
That would mean you can remove the return value and kvp object from `setUpEvents` as well, right?

Reply:
No this one wasn't a one time listener, this is on mouse move, so we need to set up the listener and keep it on until the `mousedown` event in order to track the shape path